### PR TITLE
Handle missing instance binding callbacks by finding the closest parent

### PR DESCRIPTION
**Pull request by dsnopek**
This attempts to solve a problem that was brought up on issue https://github.com/godotengine/godot-cpp/issues/1160 (although, this PR *doesn't* solve that issue).

Basically, the problem is that we assume we'll have the instance binding callbacks for any valid wrapper class. If we don't have the wrapper class, we output an error message and use `Object`.

However, in either of these cases:

- the class is a new class that was added to Godot after the GDExtension was created (so it won't know about it), or
- the developer explicitly excluded the class from godot-cpp

... we want to do what we're already doing for unexposed classes, which is to find the closest parent class that we _do_ know about.

This will only output an error message, if we're unable to find any ancestor classes that we know about.

This depends on PR https://github.com/godotengine/godot-cpp/pull/1164 which is currently a draft. So, marking this one as a draft too!

Fixes https://github.com/godotengine/godot-cpp/issues/1163

### CODE REVIEW

**Comment by DmitriySalnikov** [comment]
```
ERROR: Condition "singleton_obj == nullptr" is true. Returning: nullptr
   at: godot::ClassDBSingleton::get_singleton (.../godot_debug_draw_3d/godot-cpp/gen/src/classes/class_db_singleton.cpp:44)
```

The problem is that the class is now called `ClassDBSingleton`, and not `ClassDB`, as singleton is named.

I added this patch:
```diff
diff --git a/binding_generator.py b/binding_generator.py
index d04c698..7d6f6e7 100644
--- a/binding_generator.py
+++ b/binding_generator.py
@@ -1453,7 +1506,10 @@ def generate_engine_class_source(class_api, used_classes, fully_used_classes, us
 
     if is_singleton:
         result.append(f"{class_name} *{class_name}::get_singleton() {{")
-        result.append(f"\tconst StringName _gde_class_name = {class_name}::get_class_static();")
+        if class_name != "ClassDBSingleton":
+            result.append(f"\tconst StringName _gde_class_name = {class_name}::get_class_static();")
+        else:
+            result.append(f"\tstatic const StringName _gde_class_name = StringName(\"ClassDB\");")
         result.append(
             "\tstatic GDExtensionObjectPtr singleton_obj = internal::gdextension_interface_global_get_singleton(_gde_class_name._native_ptr());"
         )
@@ -1480,7 +1536,10 @@ def generate_engine_class_source(class_api, used_classes, fully_used_classes, us
             result.append(method_signature + " {")
 
             # Method body.
-            result.append(f"\tconst StringName _gde_class_name = {class_name}::get_class_static();")
+            if class_name != "ClassDBSingleton":
+                result.append(f"\tconst StringName _gde_class_name = {class_name}::get_class_static();")
+            else:
+                result.append(f"\tstatic const StringName _gde_class_name = StringName(\"ClassDB\");")
             result.append(f'\tconst StringName _gde_method_name = "{method["name"]}";')
             result.append(
                 f'\tstatic GDExtensionMethodBindPtr _gde_method_bind = internal::gdextension_interface_classdb_get_method_bind(_gde_class_name._native_ptr(), _gde_method_name._native_ptr(), {method["hash"]});'

```

And now everything seems to be working as before. No errors, no crashes, at first glance.
Now the difference in size is not so significant, but it is quite acceptable
37 MB -> 48 MB
image

Here is the [action](https://github.com/DmitriySalnikov/godot_debug_draw_3d/actions/runs/5487731495) in my repository with this PR. _(+9 MB for `editor` libs)_

**Comment by dsnopek** [comment]
@DmitriySalnikov:

> The problem is that the class is now called `ClassDBSingleton`, and not `ClassDB`, as singleton is named.
> 
> I added this patch:

Thanks for the patch! I've integrated it into PR https://github.com/godotengine/godot-cpp/pull/1164 and rebased this PR on that.

> And now everything seems to be working as before. No errors, no crashes, at first glance.

That's awesome! Thanks for testing it :heart:

**Comment by dsnopek** [comment]
Discussed at the GDExtension meeting, and we think this PR makes sense! However, it needs to wait on getting access to the `ClassDB` singleton's function before it can be merged.

**Comment by dsnopek** [comment]
Now that PR https://github.com/godotengine/godot-cpp/pull/1164 has been merged, I've rebased it and taken this one of draft!

We already discussed this PR at a GDExtension meeting back in July and approved it pending the merge of PR https://github.com/godotengine/godot-cpp/pull/1164, so I think it just needs passing tests, and someone to formally press the "Approve" button :-)

**Review by BastiaanOlij — APPROVED** [review]
Reviewed during GDExtension meeting, this looks like a good fix.

**Pull request merged**
